core/plugins/org.polarsys.capella.core.sirius.analysis/src/org/polarsys/capella/core/sirius/analysis/EventContextServices.java

```diff
@@ -0,0 +1,414 @@
+import static org.polarsys.capella.core.sirius.analysis.refresh.extension.InteractionRefreshExtension.isRefreshCacheEnabled;
+import static org.polarsys.capella.core.sirius.analysis.refresh.extension.InteractionRefreshExtension.putDirectEventsInCache;
+import static org.polarsys.capella.core.sirius.analysis.refresh.extension.InteractionRefreshExtension.putInstanceRoleToEventContextsCache;
+import static org.polarsys.capella.core.sirius.analysis.refresh.extension.InteractionRefreshExtension.putTimeLapseInCache;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import org.eclipse.emf.ecore.EObject;
+import org.polarsys.capella.core.data.capellacore.CapellaElement;
+import org.polarsys.capella.core.data.interaction.AbstractEnd;
+import org.polarsys.capella.core.data.interaction.AbstractFragment;
+import org.polarsys.capella.core.data.interaction.Execution;
+import org.polarsys.capella.core.data.interaction.ExecutionEnd;
+import org.polarsys.capella.core.data.interaction.InstanceRole;
+import org.polarsys.capella.core.data.interaction.InteractionFragment;
+import org.polarsys.capella.core.data.interaction.InteractionState;
+import org.polarsys.capella.core.data.interaction.MessageEnd;
+import org.polarsys.capella.core.data.interaction.Scenario;
+import org.polarsys.capella.core.data.interaction.SequenceMessage;
+import org.polarsys.capella.core.data.interaction.StateFragment;
+import org.polarsys.capella.core.data.interaction.TimeLapse;
+
+/**
+ * Compute EventContext structure for sequence diagram.
+ * 
+ * @author nlepine
+ *
+ */
+public class EventContextServices {
+
+  private EventContextServices() {
+    super();
+  }
+
+  /**
+   * Compute EventContext structure from InstanceRole if needed and put it in cache.
+   * 
+   * @param instanceRole
+   *          InstanceRole
+   * @return List<EventContext>
+   */
+  public static List<EventContext> getEventContexts(InstanceRole instanceRole) {
+    List<EventContext> eventContexts = getInstanceRoleToEventContextCache(instanceRole);
+    if (eventContexts == null) {
+      eventContexts = computeInstanceRoleEventContextStructure(instanceRole);
+      if (isRefreshCacheEnabled()) {
+        putInstanceRoleToEventContextsCache(instanceRole, eventContexts);
+      }
+    }
+    return eventContexts;
+  }
+
+  /**
+   * Get direct event on element. (InstanceRole -> Executions + StateFragment, Execution -> Execution children +
+   * StateFragment)
+   * 
+   * @param element
+   *          EObject
+   * @param instanceRole
+   *          InstanceRole
+   * @return Collection<EObject>
+   */
+  public static Collection<EObject> getDirectEvents(EObject element, InstanceRole instanceRole) {
```

Comment:
Rename `element` to `parent` and change the order of the parameters, `instanceRole` comes first, `parent` second